GraphQL: Fix weird random issue on field merge system

Moumouls opened the pull request:
### New Pull Request Checklist
<!--
    Please check the following boxes [x] before submitting your issue.
    Click the "Preview" tab for better readability.
    Thanks for contributing to Parse Server!
-->

- [ ] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [ ] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description
<!-- Add a brief description of the issue this PR solves. -->

Related issue: FILL_THIS_OUT

### Approach
<!-- Add a description of the approach in this PR. -->

### TODOs before merging
<!--
    Add TODOs that need to be completed before merging this PR.
    Delete suggested TODOs that do not apply to this PR.
-->

- [ ] Add test cases
- [ ] Add entry to changelog
- [ ] Add changes to documentation (guides, repository pages, in-code descriptions)
- [ ] Add [security check](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#security-checks)
- [ ] Add new Parse Error codes to Parse JS SDK <!-- no hard-coded error codes in Parse Server -->
- [ ] ...

mtrezza commented:
I just noticed that this PR lacks an issue. Since we classify and orient bases on issues, can you please open an issue for this?

davimacedo commented:
@Moumouls I am not sure I understood the solution. Could you please explain?

davimacedo commented:
Also, how the new test cases were failing before?

Moumouls commented:
Hi @davimacedo,

So here i just added a test to cover correctly the `ofType` condition in find last type function (before it was not covered correctly).

i think the issue come from the whole mix between auto and custom schema. The merge system used spread, reassign then a graphql tools stitch. So sometimes it seems that when a custom query/mutation resolve an extended auto graphql type (like array of users with a custom field), somethings went wrong and an error `cannot return null for User.id` is raised.

So in this PR i switched to a full reassign policy and removed the stitch schema method since everything is already merged.

After this update, the issue is gone and we don't have any more this issue in our e2e tests.

The issue was only reproducible in our e2e tests, we failed to reproduce the issue in our integration testing server suit and also i failed to reproduce the issue on the parse schema.

Here i see may be 3 sources:
- May be the issue come from the reassign/spread mix
- May be the issue come from the reassign/spread mix + stitch
- May be the issue come from stitch schema

Moumouls commented:
Update: Do not merge this one our team encountered an additional complex issue when input type is overloaded, i think in this case i can have a failing test

mtrezza commented:
@Moumouls To properly track this issue, I have opened https://github.com/parse-community/parse-server/issues/7348 for you. Please feel free to update there if the discussion goes beyond this specific PR.

Moumouls commented:
Yes thanks @mtrezza

Moumouls commented:
Okay good news i found the stacked query issue, the source come from some reassign stuff, so since graphql allow to stack multi ops into one and keep same JS ref for a shared arg (test case added) errors can occur due to some internal reassign. I added for safety a deepcopy on each input, then the GraphQL api inputs will be mutation proof on multi ops.

mtrezza commented:
@Moumouls What does that mean for this PR? Is a PR necessary or was the issue specific to your deployment?

mtrezza commented:
Changing to draft; purpose of PR is unclear, see https://github.com/parse-community/parse-server/pull/7344#issuecomment-852886161

Moumouls commented:
@mtrezza  this PR fix following issues:
- GraphQL API Introspection schema inconsistency, now correctly ordered (alphabetical order)
- GraphQL multi OPS arg isolation, now using `deepcopy` to prevent arg mutation on other ops (could lead to an unwanted query)
- Merge schema system lose some ref some times

mtrezza commented:
> ## ⚠️ Important change for merging PRs from Parse Server 5.0 onwards!
  >
  > We are planning to release the first beta version of Parse Server 5.0 in October 2021.
  >
  > If a PR contains a breaking change and is not merged before the _beta_ release of Parse Server 5.0, it cannot be merged until the end of 2022. Instead it has to follow the [Deprecation Policy](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#deprecation-policy) and phase-in breaking changes to be merged during the course of 2022.
  >
  > One of the most voiced community feedbacks was the demand for predictability in breaking changes to make it easy to upgrade Parse Server. We have made a first step towards this by introducing the Deprecation Policy in February 2021 that assists to phase-in breaking changes, giving developers time to adapt. We will follow-up with the introduction of [Release Automation](https://github.com/parse-community/parse-server/issues/7271) and a branch model that will allow breaking changes only with a new major release, scheduled for the beginning of each calendar year.
  >
  > We understand that some PRs are a long time in the making and we very much appreciate your contribution. We want to make it easy for PRs that contain a breaking change and were created _before_ the introduction of the Deprecation Policy. These PRs can be merged with a breaking change without being phased-in before the beta release of Parse Server 5.0. We are making this exception because we appreciate that this is a time of transition that requires additional effort from contributors to adapt. We encourage everyone to prepare their PRs until the end of September and account for review time and possible adaptions.
  > 
  > If a PR contains a breaking change and should be merged before the beta release, please mention @parse-community/server-maintenance and we will coordinate with you to merge the PR.
  >
  > *Thanks for your contribution and support during this transition to Parse Server release automation!*

mtrezza commented:
@Moumouls is this ready for review?

Moumouls commented:
@mtrezza  thanks for the ping, now we should be okay here, ready to review/merge !

Moumouls commented:
@mtrezza  seems that we have some issues with Postgres, if you have an idea ? https://github.com/parse-community/parse-server/pull/7344/checks?check_run_id=3845831332

parseplatformorg commented:
🎉 This change has been released in version [5.0.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/5.0.0-beta.1)

parseplatformorg commented:
🎉 This change has been released in version [5.0.0](https://github.com/parse-community/parse-server/releases/tag/5.0.0)
